pilot/pkg/xds/cds.go

```diff
@@ -69,20 +70,14 @@ func cdsNeedsPush(req *model.PushRequest, proxy *model.Proxy) bool {
 }
 
 func (c CdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource,
-	updates *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
+	updates *model.PushRequest) (model.Resources, model.DeletedResources, model.XdsLogDetails, error) {
 	if !cdsNeedsPush(updates, proxy) {
-		return nil, model.DefaultXdsLogDetails, nil
+		return nil, nil, model.DefaultXdsLogDetails, nil
 	}
-	clusters, logs := c.Server.ConfigGenerator.BuildClusters(proxy, updates)
-	return clusters, logs, nil
-}
-
-// GenerateDeltas for CDS currently only builds deltas when services change. todo implement changes for DestinationRule, etc
-func (c CdsGenerator) GenerateDeltas(proxy *model.Proxy, push *model.PushContext, updates *model.PushRequest,
-	w *model.WatchedResource) (model.Resources, []string, model.XdsLogDetails, bool, error) {
-	if !cdsNeedsPush(updates, proxy) {
-		return nil, nil, model.DefaultXdsLogDetails, false, nil
+	if c.delta {
+		updatedClusters, removedClusters, logs := c.Server.ConfigGenerator.BuildDeltaClusters(proxy, updates, w)
+		return updatedClusters, removedClusters, logs, nil
```

Comment:
If we are going this direction, I would change the signature of `Generate` to return a `bool` for whether delta generation was actually used, and return true here.

Reply:
I did not understand this part. What we care is there are removed resources or not right? Then why do we need this explicit flag ? May be I am missing some thing here